### ISSUE TITLE
Needless rebuilds when run with or without --workspace

### ISSUE THREAD

**Issue by ijackson** [rust-lang/cargo#9114]
**Problem**

cargo rebuilds things if I switch between `cargo build` and `cargo build --workspace`.

**Steps**

1. `git clone https://salsa.debian.org/iwj/otter`
2. `cd otter`
3. `git checkout c9b5a88d7a33a6e1eac22736be0a38c6815e1706`
4. `cargo build --workspace`
5. `cargo build`

To repeat, restarting from after step 4 or 5:

 3. `touch src/updates.rs`
 4. as above
 5. as above

**Expected behaviour**

The whole workspace including the toplevel package `otter` was built in step 4, so step 5 should not build anything and should be very quick.

**Actual behaviour**

cargo rebuilds `otter` (including library and binaries) in step 5 (about six seconds on my laptop in a repeated test)

**Notes**

Output of `cargo version`:
```
cargo 1.51.0-nightly (783bc43c6 2021-01-20)
```

**Comment by ijackson** [comment]
I have been attempting to work around this and the behaviours I'm seeing make me think it may be related to #4463.

**Comment by ijackson** [comment]
A workaround in my environment is to always pass `--workspace` (plus, usually, suitable `-p` options) to cargo.

**Comment by ehuss** [comment]
Yea, this is expected behavior in that the features chosen depend on which packages are being built.  Closing as a duplicate of #4463.

There are other workarounds such as a common package that forces the same features (see [rustc-workspace-hack](https://github.com/rust-lang/rust/blob/master/src/tools/rustc-workspace-hack/Cargo.toml) as an example).
